File: DataStructures/Queues/leetcode/PriorityQueues/23. Merge k Sorted Lists/Merge-k-Sorted-Lists.ts
// 以下は、TypeScript (v5.1) + Node.js 18.16.1 向けに最適化された LeetCode 用解法です。**クラスではなく関数ベース**で記述し、\*\*優先度付きキュー（最小ヒープ）\*\*を使って高速に `k` 個のソート済み連結リストをマージします。

// ---

// ### ✅ 処理概要

// * ヒープを使って `O(N log k)` 時間でマージ
// * `ListNode` 型は LeetCode 環境で事前に定義されていると仮定

// ---

// ### ✅ 時間・空間計算量

// * **時間計算量：**`O(N log k)`
// * **空間計算量：**`O(k)`（ヒープサイズ）

// ---

// ### ✅ TypeScript 実装（関数ベース）

// ```ts
// LeetCode 環境で提供されている ListNode 型定義を前提とする
// class ListNode {
//   val: number;
//   next: ListNode | null;
//   constructor(val?: number, next?: ListNode | null) {
//     this.val = (val===undefined ? 0 : val);
//     this.next = (next===undefined ? null : next);
//   }
// }

/**
 * k個のソート済み連結リストをマージして1つのソート済み連結リストを返す
 * 
 * @param lists - ListNode型の配列。各要素はソート済み連結リストの先頭ノード
 * @returns ListNode | null - マージされたソート済み連結リストの先頭ノード
 */
function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
  // 最小ヒープを表現するクラス（バイナリヒープ）
  class MinHeap {
    private heap: Array<ListNode>;

    constructor() {
      this.heap = [];
    }

    push(node: ListNode): void {
      this.heap.push(node);
      this._heapifyUp(this.heap.length - 1);
    }

    pop(): ListNode | undefined {
      if (this.heap.length === 0) return undefined;
      const min = this.heap[0];
      const end = this.heap.pop()!;
      if (this.heap.length > 0) {
        this.heap[0] = end;
        this._heapifyDown(0);
      }
      return min;
    }

    size(): number {
      return this.heap.length;
    }

    private _heapifyUp(index: number): void {
      while (index > 0) {
        const parent = Math.floor((index - 1) / 2);
        if (this.heap[index].val < this.heap[parent].val) {
          [this.heap[index], this.heap[parent]] = [this.heap[parent], this.heap[index]];
          index = parent;
        } else {
          break;
        }
      }
    }

    private _heapifyDown(index: number): void {
      const length = this.heap.length;
      while (true) {
        let smallest = index;
        const left = 2 * index + 1;
        const right = 2 * index + 2;

        if (left < length && this.heap[left].val < this.heap[smallest].val) {
          smallest = left;
        }
        if (right < length && this.heap[right].val < this.heap[smallest].val) {
          smallest = right;
        }
        if (smallest === index) break;
        [this.heap[index], this.heap[smallest]] = [this.heap[smallest], this.heap[index]];
        index = smallest;
      }
    }
  }

  const heap = new MinHeap();

  // 各リストの先頭ノードをヒープに追加
  for (const node of lists) {
    if (node !== null) {
      heap.push(node);
    }
  }

  const dummy = new ListNode(0);
  let current = dummy;

  while (heap.size() > 0) {
    const node = heap.pop()!;
    current.next = node;
    current = current.next;
    if (node.next !== null) {
      heap.push(node.next);
    }
  }

  return dummy.next;
}
// ```

// ---

// ### ✅ テスト用ユーティリティ（ローカル確認時）

// ```ts
// // 配列を連結リストに変換
// function arrayToList(arr: number[]): ListNode | null {
//   const dummy = new ListNode(0);
//   let current = dummy;
//   for (const val of arr) {
//     current.next = new ListNode(val);
//     current = current.next;
//   }
//   return dummy.next;
// }

// // 連結リストを配列に変換
// function listToArray(head: ListNode | null): number[] {
//   const result: number[] = [];
//   while (head) {
//     result.push(head.val);
//     head = head.next;
//   }
//   return result;
// }

// // 使用例
// const input = [[1,4,5], [1,3,4], [2,6]].map(arrayToList);
// const result = mergeKLists(input);
// console.log(listToArray(result)); // => [1,1,2,3,4,4,5,6]
// ```

// ---

// ### ✅ パフォーマンスの目安（LeetCode想定）

// * **実行時間：** 約 `100ms ~ 200ms`（最大入力 10⁴ ノード時）
// * **メモリ使用量：** 約 `50MB ~ 60MB`

// ---

// 必要であれば、分割統治法（Divide & Conquer）による実装も併せて提供可能です。

File: DataStructures/Queues/leetcode/PriorityQueues/23. Merge k Sorted Lists/Merge-k-Sorted-Lists.py
# 以下は、LeetCode形式に対応した **Python (CPython 3.11.4)** による `k` 個のソート済み連結リストをマージする関数です。

# ---

# ## ✅ 実装方針

# * **`heapq` モジュール**を使用して最小ヒープを構築（`(val, list_index, node)` 形式で比較可能に）
# * 各リストの先頭ノードをヒープに投入し、順次最小ノードを取り出して連結リストに追加
# * 各ノードの `.next` があれば再度ヒープに追加

# ---

# ## ✅ 計算量

# * **時間計算量:** `O(N log k)`
#   `N`: 総ノード数、`k`: リスト数
# * **空間計算量:** `O(k)`（ヒープサイズ）

# ---

# ## ✅ Python 実装（LeetCode形式）

# ```python
from typing import List, Optional
import heapq

# LeetCodeで与えられる ListNode クラス定義
class ListNode:
    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        """
        複数のソート済み連結リストを1つのソート済み連結リストに統合する関数

        :param lists: List[Optional[ListNode]]
            ソート済みの連結リストのリスト（最大k個）
        :return: Optional[ListNode]
            マージされたソート済み連結リストの先頭ノード
        """

        # Python の heapq は最小ヒープで、タプルで比較
        # ヒープの要素は (ノードの値, リストのインデックス, ノード)
        # ※リストのインデックスを入れることで、値が同じ場合も比較可能
        heap: List[tuple[int, int, ListNode]] = []

        for i, node in enumerate(lists):
            if node:
                heapq.heappush(heap, (node.val, i, node))

        dummy = ListNode(0)
        current = dummy

        while heap:
            _, i, node = heapq.heappop(heap)
            current.next = node
            current = current.next
            if node.next:
                heapq.heappush(heap, (node.next.val, i, node.next))

        return dummy.next

# ## ✅ 処理の流れ図解

# ```
# Input: [
#     1 -> 4 -> 5,
#     1 -> 3 -> 4,
#     2 -> 6
# ]

# MinHeap操作:
#   初期ヒープ: [(1, 0, node1), (1, 1, node2), (2, 2, node3)]

#   出力:
#     1 -> 1 -> 2 -> 3 -> 4 -> 4 -> 5 -> 6
# ```

# ---

# ## ✅ ヘルパー関数（ローカルでのテスト用）

# ```python
# # 配列から連結リストを作成
# def array_to_list(arr: List[int]) -> Optional[ListNode]:
#     dummy = ListNode(0)
#     current = dummy
#     for val in arr:
#         current.next = ListNode(val)
#         current = current.next
#     return dummy.next

# # 連結リストを配列に変換
# def list_to_array(head: Optional[ListNode]) -> List[int]:
#     result = []
#     while head:
#         result.append(head.val)
#         head = head.next
#     return result

# # 使用例
# if __name__ == "__main__":
#     input_lists = [
#         array_to_list([1, 4, 5]),
#         array_to_list([1, 3, 4]),
#         array_to_list([2, 6])
#     ]
#     merged = Solution().mergeKLists(input_lists)
#     print(list_to_array(merged))  # → [1, 1, 2, 3, 4, 4, 5, 6]
# ```

# ---

# ## ✅ 処理時間・メモリ消費（LeetCode参考）

# | 項目     | 規模             | 備考              |
# | ------ | -------------- | --------------- |
# | 時間計算量  | `O(N log k)`   | 全ノード数 N, リスト数 k |
# | 空間計算量  | `O(k)`         | 優先度付きキューのサイズ    |
# | 実行時間目安 | `80ms ~ 160ms` | 入力サイズにより変動      |
# | メモリ使用量 | `20MB ~ 60MB`  | 1万ノード時も効率良く処理可能 |

# ---

# ご希望であれば、\*\*分割統治法（再帰によるマージ）\*\*の実装や、**図解による解析**も併せて提供可能です。
